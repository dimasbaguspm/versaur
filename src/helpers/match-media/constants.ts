/**
 * Standard Tailwind breakpoint definitions
 * These align with common Tailwind CSS breakpoints
 */
export const BREAKPOINT_MOBILE = '(max-width: 767px)' // 0-767px
export const BREAKPOINT_TABLET = '(min-width: 768px) and (max-width: 1023px)' // 768-1023px
export const BREAKPOINT_DESKTOP = '(min-width: 1024px)' // 1024px+
