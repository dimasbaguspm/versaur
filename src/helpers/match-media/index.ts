// Base foundation
export * from './constants'
export * from './use-match-media'
export * from './match-media'

// Built-in breakpoint utilities
export * from './built-in'
