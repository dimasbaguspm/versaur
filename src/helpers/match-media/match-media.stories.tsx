import type { Meta, StoryObj } from '@storybook/react'
import { useState } from 'react'
import {
  BREAKPOINT_DESKTOP,
  BREAKPOINT_MOBILE,
  BREAKPOINT_TABLET,
  DesktopBreakpoint,
  MatchMedia,
  MobileBreakpoint,
  TabletBreakpoint,
  useDesktopBreakpoint,
  useMatchMedia,
  useMobileBreakpoint,
  useTabletBreakpoint,
} from './index'

/**
 * Match Media helpers provide responsive breakpoint detection using the native `matchMedia` API.
 *
 * This story demonstrates the various hooks and components available for responsive design.
 *
 * **Features:**
 * - Standard Tailwind-aligned breakpoint constants
 * - Hooks for programmatic breakpoint detection
 * - JSX components for declarative responsive rendering
 * - Uses native `matchMedia` API for optimal performance
 */
const meta = {
  title: 'Helpers/Match Media',
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'Responsive breakpoint utilities using native matchMedia API for performance and reliability.',
      },
    },
  },
  tags: ['autodocs'],
} satisfies Meta

export default meta
type Story = StoryObj<typeof meta>

/**
 * Demonstrates all breakpoint detection hooks in action.
 * Resize your browser window to see the hooks respond to viewport changes.
 */
const AllHooksDemo = () => {
  const isMobile = useMobileBreakpoint()
  const isTablet = useTabletBreakpoint()
  const isDesktop = useDesktopBreakpoint()

  return (
    <div className='space-y-4 p-6 rounded-lg border border-neutral'>
      <div className='space-y-2'>
        <h3 className='text-lg font-semibold text-foreground'>
          Current Breakpoints
        </h3>
        <p className='text-sm text-foreground-light'>
          Resize your browser window to see changes
        </p>
      </div>

      <div className='space-y-2'>
        <div className='flex items-center gap-2'>
          <div
            className={`w-3 h-3 rounded-full ${isMobile ? 'bg-success' : 'bg-neutral'}`}
          />
          <span className='text-sm text-foreground'>
            Mobile (0-767px): {isMobile ? '‚úì Active' : '‚úó Inactive'}
          </span>
        </div>

        <div className='flex items-center gap-2'>
          <div
            className={`w-3 h-3 rounded-full ${isTablet ? 'bg-success' : 'bg-neutral'}`}
          />
          <span className='text-sm text-foreground'>
            Tablet (768-1023px): {isTablet ? '‚úì Active' : '‚úó Inactive'}
          </span>
        </div>

        <div className='flex items-center gap-2'>
          <div
            className={`w-3 h-3 rounded-full ${isDesktop ? 'bg-success' : 'bg-neutral'}`}
          />
          <span className='text-sm text-foreground'>
            Desktop (1024px+): {isDesktop ? '‚úì Active' : '‚úó Inactive'}
          </span>
        </div>
      </div>
    </div>
  )
}

export const AllHooks: Story = {
  render: () => <AllHooksDemo />,
}

/**
 * Shows how to use the base `useMatchMedia` hook with custom media queries.
 */
const CustomMediaQueryDemo = () => {
  const isLargeScreen = useMatchMedia('(min-width: 1440px)')
  const isPortrait = useMatchMedia('(orientation: portrait)')
  const prefersDark = useMatchMedia('(prefers-color-scheme: dark)')

  return (
    <div className='space-y-4 p-6 rounded-lg border border-neutral'>
      <div className='space-y-2'>
        <h3 className='text-lg font-semibold text-foreground'>
          Custom Media Queries
        </h3>
        <p className='text-sm text-foreground-light'>
          Using useMatchMedia with custom queries
        </p>
      </div>

      <div className='space-y-2'>
        <div className='flex items-center gap-2'>
          <div
            className={`w-3 h-3 rounded-full ${isLargeScreen ? 'bg-success' : 'bg-neutral'}`}
          />
          <span className='text-sm text-foreground'>
            Large Screen (1440px+): {isLargeScreen ? '‚úì Yes' : '‚úó No'}
          </span>
        </div>

        <div className='flex items-center gap-2'>
          <div
            className={`w-3 h-3 rounded-full ${isPortrait ? 'bg-success' : 'bg-neutral'}`}
          />
          <span className='text-sm text-foreground'>
            Portrait Orientation: {isPortrait ? '‚úì Yes' : '‚úó No'}
          </span>
        </div>

        <div className='flex items-center gap-2'>
          <div
            className={`w-3 h-3 rounded-full ${prefersDark ? 'bg-success' : 'bg-neutral'}`}
          />
          <span className='text-sm text-foreground'>
            Prefers Dark Mode: {prefersDark ? '‚úì Yes' : '‚úó No'}
          </span>
        </div>
      </div>
    </div>
  )
}

export const CustomMediaQuery: Story = {
  render: () => <CustomMediaQueryDemo />,
}

/**
 * Demonstrates the generic MatchMedia component with custom queries.
 */
export const GenericMatchMedia: Story = {
  render: () => {
    return (
      <div className='space-y-4 p-6 rounded-lg border border-neutral'>
        <div className='space-y-2'>
          <h3 className='text-lg font-semibold text-foreground'>
            MatchMedia Component
          </h3>
          <p className='text-sm text-foreground-light'>
            Generic component for any media query
          </p>
        </div>

        <div className='space-y-3'>
          <MatchMedia query='(min-width: 1440px)'>
            <div className='p-4 bg-success-soft border border-success rounded-lg'>
              <p className='text-sm font-medium text-success-bold'>
                üñ•Ô∏è Large Screen (1440px+)
              </p>
              <p className='text-xs text-success mt-1'>
                Rendered with {'<MatchMedia query="(min-width: 1440px)" />'}
              </p>
            </div>
          </MatchMedia>

          <MatchMedia query='(prefers-color-scheme: dark)'>
            <div className='p-4 bg-info-soft border border-info rounded-lg'>
              <p className='text-sm font-medium text-info-bold'>
                üåô Dark Mode Preference
              </p>
              <p className='text-xs text-info mt-1'>
                Rendered with{' '}
                {'<MatchMedia query="(prefers-color-scheme: dark)" />'}
              </p>
            </div>
          </MatchMedia>

          <MatchMedia query='(orientation: portrait)'>
            <div className='p-4 bg-warning-soft border border-warning rounded-lg'>
              <p className='text-sm font-medium text-warning-bold'>
                üì± Portrait Orientation
              </p>
              <p className='text-xs text-warning mt-1'>
                Rendered with {'<MatchMedia query="(orientation: portrait)" />'}
              </p>
            </div>
          </MatchMedia>
        </div>

        <div className='mt-4 p-3 bg-info-soft border border-info-light rounded'>
          <p className='text-xs text-info-bold'>
            üí° Tip: MatchMedia works with any valid CSS media query, not just
            breakpoints!
          </p>
        </div>
      </div>
    )
  },
}

/**
 * Demonstrates the declarative breakpoint components.
 * These components conditionally render their children based on viewport size.
 */
export const BreakpointComponents: Story = {
  render: () => {
    return (
      <div className='space-y-4 p-6 rounded-lg border border-neutral'>
        <div className='space-y-2'>
          <h3 className='text-lg font-semibold text-foreground'>
            Breakpoint Components
          </h3>
          <p className='text-sm text-foreground-light'>
            Resize to see different content
          </p>
        </div>

        <div className='space-y-3'>
          <MobileBreakpoint>
            <div className='p-4 bg-primary-soft border border-primary rounded-lg'>
              <p className='text-sm font-medium text-primary-bold'>
                üì± Mobile View (0-767px)
              </p>
              <p className='text-xs text-primary mt-1'>
                This content only appears on mobile devices
              </p>
            </div>
          </MobileBreakpoint>

          <TabletBreakpoint>
            <div className='p-4 bg-secondary-soft border border-secondary rounded-lg'>
              <p className='text-sm font-medium text-secondary-bold'>
                üì± Tablet View (768-1023px)
              </p>
              <p className='text-xs text-secondary mt-1'>
                This content only appears on tablets
              </p>
            </div>
          </TabletBreakpoint>

          <DesktopBreakpoint>
            <div className='p-4 bg-tertiary-soft border border-tertiary rounded-lg'>
              <p className='text-sm font-medium text-tertiary-bold'>
                üñ•Ô∏è Desktop View (1024px+)
              </p>
              <p className='text-xs text-tertiary mt-1'>
                This content only appears on desktop
              </p>
            </div>
          </DesktopBreakpoint>
        </div>
      </div>
    )
  },
}

/**
 * Shows a practical example of using breakpoints to conditionally render navigation.
 */
const ResponsiveNavigationDemo = () => {
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false)

  return (
    <div className='w-full'>
      {/* Mobile Navigation */}
      <MobileBreakpoint>
        <div className='p-4 bg-background border border-border rounded-lg'>
          <div className='flex items-center justify-between'>
            <span className='font-semibold text-foreground'>Logo</span>
            <button
              onClick={() => setMobileMenuOpen(!mobileMenuOpen)}
              className='px-3 py-2 text-sm bg-primary text-white rounded-md'
            >
              {mobileMenuOpen ? 'Close' : 'Menu'}
            </button>
          </div>
          {mobileMenuOpen && (
            <nav className='mt-4 space-y-2'>
              <a
                href='#'
                className='block px-3 py-2 text-sm text-foreground hover:bg-neutral rounded'
              >
                Home
              </a>
              <a
                href='#'
                className='block px-3 py-2 text-sm text-foreground hover:bg-neutral rounded'
              >
                Products
              </a>
              <a
                href='#'
                className='block px-3 py-2 text-sm text-foreground hover:bg-neutral rounded'
              >
                About
              </a>
            </nav>
          )}
        </div>
      </MobileBreakpoint>

      {/* Tablet Navigation */}
      <TabletBreakpoint>
        <div className='p-4 bg-background border border-border rounded-lg'>
          <div className='flex items-center justify-between'>
            <span className='font-semibold text-foreground'>Logo</span>
            <nav className='flex gap-4'>
              <a
                href='#'
                className='text-sm text-foreground hover:text-primary'
              >
                Home
              </a>
              <a
                href='#'
                className='text-sm text-foreground hover:text-primary'
              >
                Products
              </a>
              <a
                href='#'
                className='text-sm text-foreground hover:text-primary'
              >
                About
              </a>
            </nav>
          </div>
        </div>
      </TabletBreakpoint>

      {/* Desktop Navigation */}
      <DesktopBreakpoint>
        <div className='p-4 bg-background border border-border rounded-lg'>
          <div className='flex items-center justify-between'>
            <div className='flex items-center gap-8'>
              <span className='font-semibold text-foreground'>Logo</span>
              <nav className='flex gap-6'>
                <a
                  href='#'
                  className='text-sm text-foreground hover:text-primary'
                >
                  Home
                </a>
                <a
                  href='#'
                  className='text-sm text-foreground hover:text-primary'
                >
                  Products
                </a>
                <a
                  href='#'
                  className='text-sm text-foreground hover:text-primary'
                >
                  About
                </a>
                <a
                  href='#'
                  className='text-sm text-foreground hover:text-primary'
                >
                  Contact
                </a>
              </nav>
            </div>
            <button className='px-4 py-2 text-sm bg-primary text-white rounded-md'>
              Sign In
            </button>
          </div>
        </div>
      </DesktopBreakpoint>
    </div>
  )
}

export const ResponsiveNavigation: Story = {
  render: () => <ResponsiveNavigationDemo />,
}

/**
 * Demonstrates the breakpoint constants that can be used with the base hook.
 */
export const BreakpointConstants: Story = {
  render: () => {
    return (
      <div className='space-y-4 p-6 rounded-lg border border-neutral'>
        <div className='space-y-2'>
          <h3 className='text-lg font-semibold text-foreground'>
            Breakpoint Constants
          </h3>
          <p className='text-sm text-foreground-light'>
            Standard Tailwind-aligned breakpoints
          </p>
        </div>

        <div className='space-y-3'>
          <div className='p-3 bg-neutral-soft rounded border border-neutral-light'>
            <code className='text-xs font-mono text-foreground'>
              BREAKPOINT_MOBILE = "{BREAKPOINT_MOBILE}"
            </code>
          </div>

          <div className='p-3 bg-neutral-soft rounded border border-neutral-light'>
            <code className='text-xs font-mono text-foreground'>
              BREAKPOINT_TABLET = "{BREAKPOINT_TABLET}"
            </code>
          </div>

          <div className='p-3 bg-neutral-soft rounded border border-neutral-light'>
            <code className='text-xs font-mono text-foreground'>
              BREAKPOINT_DESKTOP = "{BREAKPOINT_DESKTOP}"
            </code>
          </div>
        </div>

        <div className='mt-4 p-3 bg-info-soft border border-info-light rounded'>
          <p className='text-xs text-info-bold'>
            üí° Tip: Use these constants with useMatchMedia() for consistent
            breakpoint detection across your app.
          </p>
        </div>
      </div>
    )
  },
}
