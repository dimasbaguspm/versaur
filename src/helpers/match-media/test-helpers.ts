import { vi } from 'vitest'

/**
 * Creates a mock MediaQueryList for testing
 */
export const createMatchMediaMock = (matches: boolean) => {
  const listeners: ((event: MediaQueryListEvent) => void)[] = []

  return {
    matches,
    media: '',
    onchange: null,
    addListener: vi.fn(), // Deprecated
    removeListener: vi.fn(), // Deprecated
    addEventListener: vi.fn(
      (event: string, listener: (event: MediaQueryListEvent) => void) => {
        if (event === 'change') {
          listeners.push(listener)
        }
      }
    ),
    removeEventListener: vi.fn(
      (event: string, listener: (event: MediaQueryListEvent) => void) => {
        if (event === 'change') {
          const index = listeners.indexOf(listener)
          if (index > -1) {
            listeners.splice(index, 1)
          }
        }
      }
    ),
    dispatchEvent: vi.fn(),
    // Helper to trigger change event
    _triggerChange: (newMatches: boolean) => {
      listeners.forEach(listener => {
        listener({ matches: newMatches } as MediaQueryListEvent)
      })
    },
  }
}

/**
 * Setup and teardown for matchMedia mocking
 */
export const setupMatchMediaMock = () => {
  let originalMatchMedia: typeof window.matchMedia

  const beforeEach = () => {
    originalMatchMedia = window.matchMedia
  }

  const afterEach = () => {
    window.matchMedia = originalMatchMedia
  }

  return { beforeEach, afterEach }
}
