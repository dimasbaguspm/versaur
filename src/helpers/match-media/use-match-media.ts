import { useEffect, useState } from 'react'

/**
 * Base hook to detect if a media query matches
 * Uses the native matchMedia API for optimal performance
 *
 * @param query - CSS media query string
 * @returns boolean indicating if the query matches
 *
 * @example
 * ```tsx
 * const isMobile = useMatchMedia(BREAKPOINT_MOBILE)
 * const isDesktop = useMatchMedia(BREAKPOINT_DESKTOP)
 * const prefersDark = useMatchMedia('(prefers-color-scheme: dark)')
 * ```
 */
export function useMatchMedia(query: string): boolean {
  const [matches, setMatches] = useState<boolean>(() => {
    if (typeof window === 'undefined') return false
    return window.matchMedia(query).matches
  })

  useEffect(() => {
    if (typeof window === 'undefined') return

    const mediaQuery = window.matchMedia(query)
    const handler = (event: MediaQueryListEvent) => setMatches(event.matches)

    // Set initial value
    setMatches(mediaQuery.matches)

    // Modern browsers support addEventListener on MediaQueryList
    mediaQuery.addEventListener('change', handler)

    return () => {
      mediaQuery.removeEventListener('change', handler)
    }
  }, [query])

  return matches
}
