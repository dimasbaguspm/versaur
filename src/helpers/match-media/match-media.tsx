import { useMatchMedia } from './use-match-media'

/**
 * Props for MatchMedia component
 */
interface MatchMediaProps {
  /** CSS media query string to match against */
  query: string
  /** Content to render when query matches */
  children: React.ReactNode
}

/**
 * Component that conditionally renders children based on a media query
 *
 * @example
 * ```tsx
 * <MatchMedia query="(min-width: 1024px)">
 *   <DesktopContent />
 * </MatchMedia>
 *
 * <MatchMedia query="(prefers-color-scheme: dark)">
 *   <DarkModeUI />
 * </MatchMedia>
 * ```
 */
export function MatchMedia({ query, children }: MatchMediaProps) {
  const matches = useMatchMedia(query)
  return matches ? <>{children}</> : null
}
