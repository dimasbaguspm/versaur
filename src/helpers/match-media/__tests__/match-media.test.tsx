import { render, screen } from '@testing-library/react'
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest'
import { MatchMedia } from '../match-media'
import { createMatchMediaMock, setupMatchMediaMock } from '../test-helpers'

describe('MatchMedia', () => {
  const { beforeEach: setupBefore, afterEach: setupAfter } =
    setupMatchMediaMock()

  beforeEach(() => {
    setupBefore()
  })

  afterEach(() => {
    setupAfter()
  })

  it('should render children when query matches', () => {
    const mockMatchMedia = createMatchMediaMock(true)
    window.matchMedia = vi.fn(() => mockMatchMedia as unknown as MediaQueryList)

    render(
      <MatchMedia query='(min-width: 1024px)'>
        <div data-testid='content'>Desktop Content</div>
      </MatchMedia>
    )

    expect(screen.getByTestId('content')).toBeInTheDocument()
  })

  it('should not render children when query does not match', () => {
    const mockMatchMedia = createMatchMediaMock(false)
    window.matchMedia = vi.fn(() => mockMatchMedia as unknown as MediaQueryList)

    render(
      <MatchMedia query='(min-width: 1024px)'>
        <div data-testid='content'>Desktop Content</div>
      </MatchMedia>
    )

    expect(screen.queryByTestId('content')).not.toBeInTheDocument()
  })

  it('should work with custom media queries', () => {
    const mockMatchMedia = createMatchMediaMock(true)
    window.matchMedia = vi.fn(() => mockMatchMedia as unknown as MediaQueryList)

    render(
      <MatchMedia query='(prefers-color-scheme: dark)'>
        <div data-testid='dark-mode'>Dark Mode UI</div>
      </MatchMedia>
    )

    expect(screen.getByTestId('dark-mode')).toBeInTheDocument()
    expect(window.matchMedia).toHaveBeenCalledWith(
      '(prefers-color-scheme: dark)'
    )
  })

  it('should match snapshot when rendered', () => {
    const mockMatchMedia = createMatchMediaMock(true)
    window.matchMedia = vi.fn(() => mockMatchMedia as unknown as MediaQueryList)

    const { asFragment } = render(
      <MatchMedia query='(min-width: 1024px)'>
        <div>Content</div>
      </MatchMedia>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should not affect accessibility tree when hidden', () => {
    const mockMatchMedia = createMatchMediaMock(false)
    window.matchMedia = vi.fn(() => mockMatchMedia as unknown as MediaQueryList)

    const { container } = render(
      <MatchMedia query='(min-width: 1024px)'>
        <div role='navigation'>Navigation</div>
      </MatchMedia>
    )

    expect(
      container.querySelector('[role="navigation"]')
    ).not.toBeInTheDocument()
  })

  it('should preserve accessibility attributes when shown', () => {
    const mockMatchMedia = createMatchMediaMock(true)
    window.matchMedia = vi.fn(() => mockMatchMedia as unknown as MediaQueryList)

    render(
      <MatchMedia query='(min-width: 1024px)'>
        <nav role='navigation' aria-label='Main Navigation'>
          <button>Menu</button>
        </nav>
      </MatchMedia>
    )

    const nav = screen.getByRole('navigation', { name: 'Main Navigation' })
    expect(nav).toBeInTheDocument()
  })
})
