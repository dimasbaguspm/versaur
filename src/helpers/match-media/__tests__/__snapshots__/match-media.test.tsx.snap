// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MatchMedia > should match snapshot when rendered 1`] = `
<DocumentFragment>
  <div>
    Content
  </div>
</DocumentFragment>
`;
